src/toc.tsx

```diff
@@ -78,18 +84,21 @@ export class TableOfContents extends Widget {
       this.update,
       this
     );
-    this.update();
+    this.updateTOC(true);
   }
 
   /**
    * Handle an update request.
    */
   protected onUpdateRequest(msg: Message): void {
     // Don't bother if the TOC is not visible
-    if (!this.isVisible) {
+    /* if (!this.isVisible) {
```

Comment:
Why comment this out? I don't know that we should re-render (which is potentially expensive) unless it is visible.

Reply:
This is commented out because the auto-numbering feature in the notebook is completely depended on the TOC extension and if the TOC is not visible and user changed the headings in the document, we still need TOC to render the numberings in the notebook even if TOC is not visible.

Reply:
I think then we should have the auto-numbering logic get called elsewhere. The `onUpdateRequest` function is mostly about having the widget re-render itself. If other things are getting re-rendered, it is a bit outside of the scope of this function.

I am also a bit uncomfortable with the DOM replacement for the auto-numbering of documents. I think it might be less surprising to the user to actually add the numbers to the text of the document.

Reply:
Yeah because I'm just trying to mimic the behavior of toc2 extension for Jupyter Notebook (which add the rendered numbering to DOM), but I can definitely switch to adding numberings to the original document.

My question for adding numberings to the original document is, if user has some numberings manually before, do we need to detect these numberings, or we just add the auto-generated numbering to the document anyway?

Reply:
That's a good question. I'm not completely sold on whether it should change the numbering in the text, or change it in the view. Did you have any design discussions on this topic? I found this behavior kind of confusing, but it could be kind of error prone to change the actual text...

Reply:
Yeah this is my major concern at the beginning, we didn't go through design discussion because after I proposed this concern we just decided to look at what toc2 in Jupyter Notebook did and did the similar design, but I can definitely talk about this with my team.

Reply:
I'd like to think of some way that we can update the headings in the notebook DOM without actually doing it in the `onUpdateRequest` function here. The purpose of this function (it seems to me) is for the TOC to update its own content. If it is reaching out into other parts of the DOM, I think it should be elsewhere. Let's talk this week about this in person, if we can.